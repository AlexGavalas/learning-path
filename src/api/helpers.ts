import axios from 'axios';

export const fetchFileFromStorage = async (
    filename: string,
): Promise<string> => {
    const fileServerUrl = String(process.env.PUBLIC_FILE_SERVER_URL);

    const response = await axios.get(`${fileServerUrl}/${filename}`, {
        responseType: 'text',
    });

    return response.data as string;
};
