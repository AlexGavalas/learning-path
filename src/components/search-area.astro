---
import { QUERY_FIELD_NAME } from '~constants';
import Loader from '~components/loader.astro';
import Button from '~components/button.astro';
import Input from './input.astro';

const { q } = Astro.props;
---

<search-area>
    <form class="relative flex h-16 items-center">
        <Input
            label="Search notes"
            name={QUERY_FIELD_NAME}
            autocomplete="off"
            placeholder="Type here"
            value={q}
        />
        <div class="absolute bottom-0 right-0 flex h-1/2 gap-2 p-1 text-sm">
            <Button>Search</Button>
        </div>
        <div id="loader" class="hidden absolute -bottom-1 h-[2px] w-full">
            <Loader />
        </div>
    </form>
</search-area>

<script>
    import { navigate } from 'astro:transitions/client';
    import { QUERY_FIELD_NAME } from '~constants';

    class SearchArea extends HTMLElement {
        formElement: HTMLFormElement | null = null;
        inputElement: HTMLInputElement | null = null;
        loaderElement: HTMLDivElement | null = null;

        constructor() {
            super();

            this.formElement = this.querySelector('form');
            this.inputElement = this.querySelector('input');
            this.loaderElement = this.querySelector('#loader');
        }

        handleKeyUp = (e: KeyboardEvent) => {
            if (e.key === '/') {
                this.inputElement?.focus();
            }
        };

        handleSearch = (e: SubmitEvent) => {
            this.loaderElement?.classList.remove('hidden');

            const url = new URL(location.href);
            const query = this.inputElement?.value ?? '';

            if (query.length > 0) {
                url.searchParams.set(QUERY_FIELD_NAME, query);
            } else {
                url.searchParams.delete(QUERY_FIELD_NAME);
            }

            // window.location.href = url.toString();
            navigate(url.toString());

            e.preventDefault();
        };

        connectedCallback() {
            document.addEventListener('keyup', this.handleKeyUp);
            this.formElement?.addEventListener('submit', this.handleSearch);
        }

        disconnectedCallback() {
            document.removeEventListener('keyup', this.handleKeyUp);
            this.formElement?.removeEventListener('submit', this.handleSearch);
        }
    }

    customElements.define('search-area', SearchArea);
</script>
