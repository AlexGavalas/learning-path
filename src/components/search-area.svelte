<script lang="ts">
    import { onDestroy, onMount } from 'svelte';

    import Button from '~components/button.svelte';
    import Input from '~components/input.svelte';
    import Loader from '~components/loader.svelte';
    import { QUERY_FIELD_NAME } from '~constants';

    interface $$Props {
        q: string;
    }

    export let q: $$Props['q'];

    let inputElement: any;
    let query = q;
    let loading = false;

    const keyPressHandler = (e: KeyboardEvent) => {
        if (e.key === '/') {
            e.stopImmediatePropagation();
            inputElement.focus();
        }
    };

    onMount(() => {
        typeof document !== 'undefined' &&
            document.addEventListener('keyup', keyPressHandler);
    });

    onDestroy(() => {
        typeof document !== 'undefined' &&
            document.removeEventListener('keyup', keyPressHandler);
    });

    const onSubmit = (e: SubmitEvent) => {
        e.preventDefault();

        loading = true;

        const url = new URL(location.href);

        if (query.length > 0) {
            url.searchParams.set(QUERY_FIELD_NAME, query);
        } else {
            url.searchParams.delete(QUERY_FIELD_NAME);
        }

        window.location.href = url.toString();
    };
</script>

<form on:submit={onSubmit} class="relative flex h-16 items-center">
    <Input
        bind:this={inputElement}
        bind:value={query}
        label="Search notes"
        name={QUERY_FIELD_NAME}
        autocomplete="off"
        placeholder="Type here"
    />
    <div class="absolute bottom-0 right-0 flex h-1/2 gap-2 p-1 text-sm">
        <Button>Search</Button>
    </div>
    {#if loading}
        <Loader />
    {/if}
</form>
