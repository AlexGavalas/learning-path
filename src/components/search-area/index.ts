export { SearchArea } from './search-area';
