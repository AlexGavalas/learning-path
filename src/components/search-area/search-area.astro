---
import { QUERY_FIELD_NAME } from '~constants';
import Loader from '~components/loader/linear.astro';
import Button from '~components/button/button.astro';
import Input from '~components/input/input.astro';
import cn from './search-area.module.css';

interface Props {
    query?: string;
}

const { query } = Astro.props;
---

<search-area>
    <form class={cn.container}>
        <Input
            autocomplete="off"
            class={cn.input}
            id={QUERY_FIELD_NAME}
            label="Search the notes"
            name={QUERY_FIELD_NAME}
            placeholder="Javascript, CSS grid ..."
            value={query}
        />
        <Button class={cn.button}>Search</Button>
        <div id="loader" class={cn.loader} hidden>
            <Loader />
        </div>
    </form>
</search-area>

<script>
    import { QUERY_FIELD_NAME } from '~constants';

    class SearchArea extends HTMLElement {
        formElement: HTMLFormElement | null = null;
        inputElement: HTMLInputElement | null = null;
        loaderElement: HTMLDivElement | null = null;

        constructor() {
            super();

            this.formElement = this.querySelector('form');
            this.inputElement = this.querySelector('input');
            this.loaderElement = this.querySelector('#loader');
        }

        handleKeyUp = (e: KeyboardEvent) => {
            if (e.key === '/') {
                this.inputElement?.focus();
            }
        };

        handleSearch = (e: SubmitEvent) => {
            if (this.loaderElement) {
                this.loaderElement.hidden = false;
            }

            const url = new URL(location.href);
            const query = this.inputElement?.value ?? '';

            if (query.length > 0) {
                url.searchParams.set(QUERY_FIELD_NAME, query);
            } else {
                url.searchParams.delete(QUERY_FIELD_NAME);
            }

            window.addEventListener(
                'pagehide',
                () => {
                    if (this.loaderElement) {
                        this.loaderElement.hidden = true;
                    }
                },
                { once: true },
            );

            window.location.href = url.toString();

            e.preventDefault();
        };

        connectedCallback() {
            const inputValue = this.inputElement?.value;

            if (inputValue) {
                this.inputElement?.setSelectionRange(
                    inputValue.length,
                    inputValue.length,
                );

                this.inputElement?.focus();
            }

            document.addEventListener('keyup', this.handleKeyUp);
            this.formElement?.addEventListener('submit', this.handleSearch);
        }

        disconnectedCallback() {
            document.removeEventListener('keyup', this.handleKeyUp);
            this.formElement?.removeEventListener('submit', this.handleSearch);
        }
    }

    customElements.define('search-area', SearchArea);
</script>
