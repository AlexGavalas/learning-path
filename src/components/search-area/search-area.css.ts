import { style } from '@vanilla-extract/css';

import { vars } from '../../styles/theme.css';

export const container = style({
    position: 'relative',
    display: 'grid',
    gap: vars.spacing.sm,
    gridTemplateColumns: '1fr auto',
    gridTemplateAreas: `"input ." "input button"`,
    '@media': {
        'screen and (max-width: 360px)': {
            gridTemplateColumns: '1fr',
        },
    },
});

export const input = style({
    display: 'grid',
    gap: vars.spacing.sm,
    gridArea: 'input',
    gridTemplateRows: 'subgrid',
});

export const searchButton = style({
    padding: `${vars.spacing.xs} ${vars.spacing.md} !important`,
    fontSize: '0.9em',
    gridArea: 'button',
});

export const loader = style({
    position: 'absolute',
    bottom: '-8px',
    height: '2px',
    width: '100%',
});
