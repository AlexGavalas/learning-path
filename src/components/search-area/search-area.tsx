import {
    type FormEventHandler,
    useCallback,
    useEffect,
    useRef,
    useState,
} from 'react';

import { Button } from '~components/button';
import { Input } from '~components/input';
import { Loader } from '~components/loader';
import { QUERY_FIELD_NAME } from '~constants';
import { useKeypress } from '~hooks/use-keypress';

export const SearchArea = ({ q }: { q: string }): JSX.Element => {
    const queryEl = useRef<HTMLInputElement | null>(null);

    const [query, setQuery] = useState(q);
    const [loading, setLoading] = useState(false);

    const keyPressHandler = useCallback((e: KeyboardEvent) => {
        e.stopImmediatePropagation();
        queryEl.current?.focus();
    }, []);

    useEffect(() => {
        setLoading(false);
    }, [q]);

    useKeypress('/', keyPressHandler);

    const onSubmit: FormEventHandler<HTMLFormElement> = (e) => {
        e.preventDefault();

        setLoading(true);

        const url = new URL(location.href);

        if (query.length > 0) {
            url.searchParams.set(QUERY_FIELD_NAME, query);
        } else {
            url.searchParams.delete(QUERY_FIELD_NAME);
        }

        window.location.href = url.toString();
    };

    return (
        <form onSubmit={onSubmit} className="relative flex h-16 items-center">
            <Input
                label="Search notes"
                name={QUERY_FIELD_NAME}
                ref={queryEl}
                value={query}
                onChange={(e) => {
                    setQuery(e.target.value);
                }}
                autoFocus={query.length > 0}
                autoComplete="off"
                placeholder="Type here"
            />
            <div className="absolute bottom-0 right-0 flex h-1/2 gap-2 p-1 text-sm">
                <Button>Search</Button>
            </div>
            {loading && <Loader />}
        </form>
    );
};
