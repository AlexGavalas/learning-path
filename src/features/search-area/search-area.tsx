import {
    Dispatch,
    FormEventHandler,
    MouseEventHandler,
    SetStateAction,
    useCallback,
} from 'react';

import { useState, useRef } from 'react';

import type { Post } from '~lib/posts';

import { supabase } from '~lib/supabase';
import { Button } from '~components/button';
import { Input } from '~components/input';
import { Loader } from '~components/loader';
import { useKeypress } from '~hooks/use-keypress';

const QUERY_FIELD_NAME = 'query';

type SearchAreaProps = {
    posts: Post[];
    setPosts: Dispatch<SetStateAction<Post[]>>;
    setLines: Dispatch<SetStateAction<Record<string, string[]>>>;
};

export const SearchArea = ({ posts, setPosts, setLines }: SearchAreaProps) => {
    const queryEl = useRef<HTMLInputElement>(null);
    const [query, setQuery] = useState('');
    const [loading, setLoading] = useState(false);

    const keyPressHandler = useCallback((e: KeyboardEvent) => {
        e.stopImmediatePropagation();
        queryEl.current?.focus();
    }, []);

    useKeypress('/', keyPressHandler);

    const onSubmit: FormEventHandler<HTMLFormElement> = async (e) => {
        e.preventDefault();

        if (!query) {
            setLines({});
            setPosts(posts);

            return;
        }

        setLoading(true);

        try {
            const { data } = await supabase.rpc('search_notes', { q: query });

            const lines =
                data?.reduce<Record<string, string[]>>(
                    (acc, { title, line }) => {
                        acc[title] = (acc[title] || []).concat(line);
                        return acc;
                    },
                    {},
                ) ?? {};

            const postTitles = data?.map(({ title }) => title) ?? [];

            const filteredPosts = posts.filter(({ title }) =>
                postTitles.includes(title),
            );

            setPosts(filteredPosts);
            setLines(lines);
        } catch (e) {
            console.error(e);
        } finally {
            setLoading(false);
        }
    };

    const onClear: MouseEventHandler<HTMLButtonElement> = () => {
        setQuery('');
        setPosts(posts);
        setLines({});
    };

    return (
        <form onSubmit={onSubmit} className="relative h-16 flex items-center">
            <Input
                label="Search notes"
                name={QUERY_FIELD_NAME}
                ref={queryEl}
                value={query}
                onChange={(e) => setQuery(e.target.value)}
                autoComplete="off"
                placeholder="Type here"
            />
            <div className="absolute right-0 bottom-0 h-1/2 flex gap-2 p-1">
                {query && (
                    <Button onClick={onClear} type="button">
                        Clear
                    </Button>
                )}
                <Button>Search</Button>
            </div>
            {loading && <Loader />}
        </form>
    );
};
