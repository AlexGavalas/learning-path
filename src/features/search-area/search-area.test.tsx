import { axe } from 'jest-axe';

import { renderWithUser, screen } from '~test/helpers';

import { SearchArea } from './search-area';

const mockPush = jest.fn();

jest.mock('next/router', () => ({
    useRouter() {
        return {
            route: '',
            pathname: '',
            query: '',
            asPath: '',
            push: mockPush,
            events: {
                on: jest.fn(),
                off: jest.fn(),
            },
            beforePopState: jest.fn(() => null),
            prefetch: jest.fn(() => null),
        };
    },
}));

const renderSearchArea = () => renderWithUser(<SearchArea />);

const baseURL = new URL('http://localhost/');

describe('<SearchArea />', () => {
    it('renders', () => {
        const { container } = renderSearchArea();

        expect(container).toMatchSnapshot();
    });

    it('is accessible', async () => {
        const { container } = renderSearchArea();

        const a11yResults = await axe(container);

        expect(a11yResults).toHaveNoViolations();
    });

    describe('when search input is empty', () => {
        it('clear button is hidden', () => {
            renderSearchArea();

            const clearButton = screen.queryByText(/clear/i);

            expect(clearButton).not.toBeInTheDocument();
        });
    });

    describe('when user presses "/"', () => {
        it('search input receives focus', async () => {
            const { user } = renderSearchArea();

            expect(screen.getByRole('textbox')).not.toHaveFocus();

            await user.keyboard('/');

            expect(screen.getByRole('textbox')).toHaveFocus();
        });
    });

    describe('when user types in search input', () => {
        it('can clear the input', async () => {
            const { user } = renderSearchArea();

            const searchInput = screen.getByRole('textbox');

            expect(searchInput).toHaveValue('');

            await user.keyboard('/abc');

            expect(searchInput).toHaveValue('abc');

            await user.click(screen.getByText(/clear/i));

            expect(searchInput).toHaveValue('');

            const url = new URL(baseURL);
            url.searchParams.delete('q');

            expect(mockPush).toHaveBeenCalledTimes(1);
            expect(mockPush).toHaveBeenCalledWith(url);
        });

        describe('when user clicks search', () => {
            it('does nothing when the search query is empty', async () => {
                const { user } = renderSearchArea();

                await user.click(
                    screen.getByRole('button', { name: /search/i }),
                );

                expect(mockPush).toHaveBeenCalledTimes(1);
                expect(mockPush).toHaveBeenCalledWith(baseURL);
            });

            it('calls the supabase rpc when it has a query', async () => {
                const { user } = renderSearchArea();
                const searchQuery = 'abc';

                await user.keyboard(`/${searchQuery}`);

                await user.click(
                    screen.getByRole('button', { name: /search/i }),
                );

                const url = new URL(baseURL);
                url.searchParams.set('q', searchQuery);

                expect(mockPush).toHaveBeenCalledTimes(1);
                expect(mockPush).toHaveBeenCalledWith(url);
            });
        });
    });
});
