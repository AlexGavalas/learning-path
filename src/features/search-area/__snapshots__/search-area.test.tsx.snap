// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchArea /> renders 1`] = `
<div>
  <form
    class="relative h-16 flex items-center"
  >
    <label
      class="w-full h-full flex flex-col"
    >
      <span>
        Search notes
      </span>
      <input
        autocomplete="off"
        class="pl-2 w-full h-1/2 border-[1px] dark:border-0 border-solid bg-transparent dark:bg-neutral-800 rounded dark:placeholder-gray-300 dark:placeholder-opacity-50"
        name="query"
        placeholder="Type here"
        value=""
      />
    </label>
    <div
      class="absolute right-0 bottom-0 h-1/2 flex gap-2 p-1"
    >
      <button
        class="p-2 flex items-center rounded border-0 text-black dark:text-white bg-gray-200 hover:bg-gray-300 dark:bg-gray-500 dark:hover:bg-gray-600"
      >
        Search
      </button>
    </div>
  </form>
</div>
`;
